pkg/client/client.go

```diff
@@ -0,0 +1,410 @@
+		}
+	}()
+}
+
+// startActions starts the go rountines to send and receive actions
+func (c *Client) startActions() {
+	c.wg.Add(1)
+
+	// results our held outside of the retry loop, because on re-connect
+	// we still want to send the responses that either failed or haven't been
+	// sent back to the agent.
+	actionResults := make(chan *proto.ActionResponse, 100)
+	go func() {
+		defer c.wg.Done()
+		for {
+			select {
+			case <-c.ctx.Done():
+				// stopped
+				return
+			default:
+			}
+
+			actionsClient, err := c.client.Actions(c.ctx)
+			if err != nil {
+				c.impl.OnError(err)
+				continue
+			}
+
+			// initial connection of stream must send the token so
+			// the Elastic Agent knows this clients token.
+			actionResults <- &proto.ActionResponse{
+				Token:  c.token,
+				Id:     "init",
+				Status: proto.ActionResponse_SUCCESS,
+				Result: []byte("{}"),
+			}
+
+			var wg sync.WaitGroup
+			wg.Add(2)
+			done := make(chan bool)
+
+			// action requests
+			go func() {
+				defer wg.Done()
+				for {
+					action, err := actionsClient.Recv()
+					if err != nil {
+						if err != io.EOF {
+							c.impl.OnError(err)
+						}
+						close(done)
+						return
+					}
+
+					actionImpl, ok := c.actions[action.Name]
+					if !ok {
+						actionResults <- &proto.ActionResponse{
+							Token:  c.token,
+							Id:     action.Id,
+							Status: proto.ActionResponse_FAILED,
+							Result: ActionErrUndefined,
+						}
+						continue
+					}
+
+					var params map[string]interface{}
+					err = json.Unmarshal(action.Params, &params)
+					if err != nil {
+						actionResults <- &proto.ActionResponse{
+							Token:  c.token,
+							Id:     action.Id,
+							Status: proto.ActionResponse_FAILED,
+							Result: ActionErrUnmarshableParams,
+						}
+						continue
+					}
+
+					// perform the action
+					go func() {
+						res, err := actionImpl.Execute(params)
+						if err != nil {
+							actionResults <- &proto.ActionResponse{
+								Token:  c.token,
+								Id:     action.Id,
+								Status: proto.ActionResponse_FAILED,
+								Result: utils.JSONMustMarshal(map[string]string{
+									"error": err.Error(),
+								}),
+							}
+							return
+						}
+						resBytes, err := json.Marshal(res)
+						if err != nil {
+							// client-side error, should have been marshal-able
+							c.impl.OnError(err)
+							actionResults <- &proto.ActionResponse{
+								Token:  c.token,
+								Id:     action.Id,
+								Status: proto.ActionResponse_FAILED,
+								Result: ActionErrUnmarshableResult,
+							}
+							return
+						}
+						actionResults <- &proto.ActionResponse{
+							Token:  c.token,
+							Id:     action.Id,
+							Status: proto.ActionResponse_SUCCESS,
+							Result: resBytes,
+						}
+					}()
+				}
+			}()
+
+			// action responses
+			go func() {
+				defer wg.Done()
+				for {
+					select {
+					case <-done:
+						return
+					case res := <-actionResults:
+						err := actionsClient.Send(res)
```

Comment:
as with checking if this fails should we return goroutines and let loop to recreate stream?

Reply:
Yes, good idea! I have updated it to disconnect and reconnect.